workflow-manager/main.go

```diff
@@ -58,38 +58,45 @@ type batchPath struct {
 	dateComponents []string
 	ID             string
 	time           time.Time
+	metadata       bool
+	avro           bool
+	sig            bool
 }
 
-func newBatchPath(batchName string) (batchPath, error) {
+func newBatchPath(batchName string) (*batchPath, error) {
```

Comment:
I thought it was Go convention to return a value from a `new...` function instead of a pointer. I guess it doesn't matter in a GC language, and I think your Go is much sharper than mine.

Reply:
This is actually a weird thing about Go that bites everybody new to the language. I was very confused about it for years.

Basically when you write a type `Foo`, you get to choose whether to pass it around as a pointer or a value. You express this choice mostly indirectly - by defining your methods with pointer receivers vs value receivers. If your methods have pointer receivers, your `newFoo` should return a pointer; if they have value receivers, your `newFoo` should return a value.

This is somewhat masked by the fact that Go does automatic dereferences: https://golang.org/ref/spec#Selectors.

So how do you choose whether Foo should be manipulated as a pointer vs a value? One simple rule of thumb: If the type is smaller than two pointers, you can save some heap allocations and generally be faster by making it a value. If it's bigger, go ahead and use a pointer.

The other upside of using a pointer, which I used here, is that in all the error cases, you can return `nil` instead of `foo{}`, which is tidier and easier to type.